signer-service/src/api/services/sep10.service.js

```diff
@@ -0,0 +1,95 @@
+
+  const { signingKey: anchorSigningKey } = await fetchTomlValues(TOKEN_CONFIG[outToken].tomlFileUrl);
+
+  const transactionSigned = new TransactionBuilder.fromXDR(challengeXDR, NETWORK_PASSPHRASE);
+  if (transactionSigned.source !== anchorSigningKey) {
+    throw new Error(`Invalid source account: ${transactionSigned.source}`);
+  }
+  if (transactionSigned.sequence !== '0') {
+    throw new Error(`Invalid sequence number: ${transactionSigned.sequence}`);
+  }
+
+  // See https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#success
+  // memo field should be empty as we assume (in this implementation) that we use the ephemeral
+  // to authenticate. But no memo sub account derivation.
+  if (transactionSigned.memo.value === '') {
+    throw new Error('Memo does not match');
+  }
+
+  const { operations } = transactionSigned;
+  // Verify the first manage_data operation
+  const firstOp = operations[0];
+  if (firstOp.type !== 'manageData') {
+    throw new Error('The first operation should be manageData');
+  }
+
+  // Only authorize a session that corresponds with the ephemeral client account
+  if (firstOp.source !== clientPublicKey) {
+    throw new Error('First manageData operation must have the client account as the source');
+  }
+
+  const { anchorExpectedKey: expectedKey, clientDomainEnabled } = TOKEN_CONFIG[outToken];
+  if (firstOp.name !== expectedKey) {
+    throw new Error(`First manageData operation should have key '${expectedKey}'`);
+  }
+  if (!firstOp.value || firstOp.value.length !== 64) {
+    throw new Error('First manageData operation should have a 64-byte random nonce as value');
+  }
+
+  // Flags to check presence of required operations
+  let hasWebAuthDomain = false;
+  let hasClientDomain = false;
+
+  // Verify extra manage_data operations, web_auth and proper client domain.
+  for (let i = 1; i < operations.length; i++) {
+    const op = operations[i];
+
+    if (op.type !== 'manageData') {
+      throw new Error('All operations should be manage_data operations');
+    }
+
+    // Verify web_auth_domain operation
+    if (op.name === 'web_auth_domain') {
+      hasWebAuthDomain = true;
+      if (op.source !== anchorSigningKey) {
+        throw new Error('web_auth_domain manage_data operation must have the server account as the source');
+      }
+    }
+
+    if (op.name === 'client_domain') {
+      hasClientDomain = true;
+      if (op.source !== clientDomainStellarKeypair.publicKey()) {
+        throw new Error('client_domain manage_data operation must have the client domain account as the source');
+      }
+    }
+  }
+
+  //  the web_auth_domain and client_domain operation must be present
+  if (!hasWebAuthDomain) {
+    throw new Error('Transaction must contain a web_auth_domain manageData operation');
+  }
+  if (!hasClientDomain && clientDomainEnabled) {
+    throw new Error('Transaction must contain a client_domain manageData operation');
+  }
```

Comment:
Should we alleviate this restriction as it's marked as optional in the protocol?

Reply:
It's mostly for us, and check what the UI client is doing. We can alleviate it, but the intention is that if the UI asks for a challenge corresponding to a specific token/anchor, let's say `eurc`, we check that we allow signing `client_domain` proof for it. Only if we allow it, and the operation correspond to what is expected, we sign.

It may be a bit of too defensive and redundant. We can analyze removing it if it doesn't add value. But it's not too much code.

Reply:
I should mention that the `client_domain` operation is defined by the initial GET request to `sep-10`, if it is defined there, it will appear in the operations.
And right now we (the backend) don't control this.

Reply:
Alright, well I guess if we always want to use the `client_domain` we can just keep the check.

Reply:
This is also useful, I think, when we want to also disable some anchor. Yes I also understand we always want to send the `client_domain`. It's defensive, but for example right now we are experiencing a failure if we specify the `client_domain` with Anclap. So we disable it here and we can be sure it's not a misconfiguration on the frontend.